File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def build_map(self, vals):
        index_map = {}
        for i in range(len(vals)):
            index_map[vals[i]] = i

        return index_map


    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:

        index_map = self.build_map(inorder)
        print(index_map)
        
        return self._buildTree(preorder, inorder, index_map)
    
    def _buildTree(self, preorder: List[int], inorder: List[int], index_map):
        if not preorder:
            return None

        if len(preorder) == 1:
            return TreeNode(val=preorder[0])

        
        root = TreeNode(val=preorder[0])
        in_root_index = inorder.index(root.val)

        root.left = self._buildTree(preorder[1:in_root_index+1], inorder[:in_root_index], index_map)
        root.right = self._buildTree(preorder[in_root_index+1:], inorder[in_root_index+1:], index_map)

        return root

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
unordered_map<int, int> inorderValueToIndex;

public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        for (int i = 0; i < inorder.size(); i++) {
            inorderValueToIndex[inorder[i]] = i;
        }

       return buildBinaryTree(preorder, 0, preorder.size(), inorder, 0, inorder.size());
    }

    TreeNode *buildBinaryTree(vector<int>& preorder, int preorderStart, int preorderEnd, vector<int>& inorder, int inorderStart, int inorderEnd) {
        if (preorderStart >= preorderEnd) {
            return nullptr;
        }

        auto root = new TreeNode(preorder[preorderStart]);

        int rootInorderIndex = inorderValueToIndex[preorder[preorderStart]];

        int leftSize = rootInorderIndex - inorderStart;
        int leftStart = preorderStart + 1;
        root->left = buildBinaryTree(preorder, leftStart, leftStart + leftSize, inorder, inorderStart, inorderStart + leftSize);
        root->right = buildBinaryTree(preorder, leftStart + leftSize, preorderEnd, inorder, inorderStart + leftSize + 1, inorderEnd);

        return root;
    }
};
